src/umpire/util/MemoryMap.hpp

```diff
@@ -0,0 +1,141 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory
+//
+// Created by David Beckingsale, david@llnl.gov
+// LLNL-CODE-747640
+//
+// All rights reserved.
+//
+// This file is part of Umpire.
+//
+// For details, see https://github.com/LLNL/Umpire
+// Please also see the LICENSE file for MIT license.
+//////////////////////////////////////////////////////////////////////////////
+#ifndef UMPIRE_MemoryMap_HPP
+#define UMPIRE_MemoryMap_HPP
+
+#include "umpire/tpl/judy/judy.h"
+
+#include "umpire/util/FixedMallocPool.hpp"
+
+#include <cstdint>
+#include <iterator>
+#include <utility>
+#include <type_traits>
+
+namespace umpire {
+namespace util {
+
+struct iterator_begin {};
+struct iterator_end {};
+
+// MemoryMap maps addresses to a templated type Value, using a
+// FixedMallocPool underneath for speed. It is not threadsafe.
+
+template <typename V>
+class MemoryMap
+{
+public:
+  using Key = void*;
+  using Value = V;
+  using KeyValuePair = std::pair<Key, Value*>;
+
+  template <bool Const = false>
+  class Iterator_ : public std::iterator<std::forward_iterator_tag, Value> {
```

Comment:
`MemoryMap` includes both const and non-const iterator implementations. `RecordList` and `AllocationMap` only have const iterators for simplicity (and because they aren't needed at the moment).